### PR TITLE
Update oocodegen.map_type to use optional slot.required

### PR DESCRIPTION
**Pull request by julesjacobsen**
Update javagen.py to check slot.required before deciding to use a primitive or wrapper type for int, float, double.

### CODE REVIEW

File: linkml/generators/oocodegen.py [hunk 1]
```diff
@@ -182,6 +186,9 @@ def create_documents(self) -> List[OODocument]:
                 elif range == "String":
                     default_value = '""'
 
+                # TODO:
```
Comment (julesjacobsen) [inline linkml/generators/oocodegen.py:189]: Needs wiring-up here! Not sure how the existing types need to be retained/overridden.

**Comment by julesjacobsen** [comment]
Still a bit to do - requires wiring-up of default_value_for_type in oocodegen (see comment in files changed)
pydanticgen needs to implement  default_value_for_type  too?

**Comment by julesjacobsen** [comment]
This has been updated to add in new Java functionality to create a Java Interface for mixin types and add in the description to the generated classes e.g.

```yaml
classes:
  annotatable:
    description: mixin for classes that support annotations
    mixin: true
    slots:
      - annotations

  annotation:
    description: a tag/value pair with the semantics of OWL Annotation
    is_a: extension
    mixins:
      - annotatable
    slots:
      - annotations
```

produces

```java

/**
 * mixin for classes that support annotations
 */
public interface Annotatable {

    public List<Annotation> annotations();
}

/**
 * a tag/value pair with the semantics of OWL Annotation
 */
public record Annotation(List<Annotation> annotations, String extensionTag, String extensionValue, List<Extension> extensions) implements Annotatable {}
```

**Comment by cmungall** [comment]
still to address:

```
Traceback (most recent call last):
  File "/home/runner/work/linkml/linkml/tests/test_generators/test_javagen.py", line 44, in test_javagen_classes
    gen.serialize(directory=JAVA_DIR)
  File "/home/runner/work/linkml/linkml/linkml/generators/javagen.py", line 126, in serialize
    code = template_interface.render(
UnboundLocalError: local variable 'template_interface' referenced before assignment
```

**Review by kevinschaper — APPROVED** [review]

**Review by cmungall — CHANGES_REQUESTED** [review]
can merge as soon as tests pass..

**Pull request merged**